### PR TITLE
Save slicing HTTP 2 headers & data

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

DefaultHttp2FrameWriter's always create aggregate promises and slices out of headers and data: both could be saved while reducing the amount of pipeline traversals in case the additional cost of creating a sliced buffer surpass the data to be written.

Modifications:

Small header and data could be copied directly in a single buffer, without any need to create aggregate promises.

Result:

Faster small data's writes

### CODE REVIEW

**Comment by franz1981** [comment]
@normanmaurer I'm not quite sure i could reuse the promise, so please take a look if I'm not assuming anything wrong...

The cutoff value has been decided by summing the required bytes for both a new sliced buffer (around 40 bytes ) + outbound buffer's entries (64  or 64 * 2 bytes, depending if headers/data paths) + promise aggregator (> 64 bytes).

**Comment by franz1981** [comment]
This is just a preliminary example of the improvement, given that it doesn't account for the saving of the aggregate promise (which translate in less morphism of call-sites around the promises handling):

before:
```
Benchmark                                (padding)  (payloadSize)  (pooled)  Mode  Cnt    Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter          0             64      true  avgt   10  112.406 ± 1.096  ns/op
Http2FrameWriterDataBenchmark.newWriter          0             64     false  avgt   10  110.992 ± 0.120  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024      true  avgt   10  102.480 ± 1.899  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024     false  avgt   10  112.736 ± 5.971  ns/op
```
now:
```

Benchmark                                (padding)  (payloadSize)  (pooled)  Mode  Cnt    Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter          0             64      true  avgt   10   91.973 ± 0.343  ns/op
Http2FrameWriterDataBenchmark.newWriter          0             64     false  avgt   10  100.445 ± 0.238  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024      true  avgt   10   94.347 ± 0.157  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024     false  avgt   10   98.489 ± 0.196  ns/op
```

which is already ~10-20% improvement just on headers/data writing.

**Comment by franz1981** [comment]
An additional benefit (specific to Vertx, but can really happen regardless), is that the data written by users could uses heap-based `ByteBuf` meaning that embedding their content into the direct buffer to send, would avoid performing the copy again before sending them to the wire, see

image

this code path just disappear with this PR, given that no heap buffer is passed.
I'm thinking to ignore the cutoff limits in case the data is heap-based, wdyt?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 1]
```diff
@@ -168,16 +174,43 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
...
+                    ctx.write(lastFrame, promiseAggregator.newPromise());
+                } else {
+                    // let's use a single buffer to reduce the number of pipeline traversals, eventually simplifying
+                    // the promise aggregation as much as possible
+                    ByteBuf frameHeader2AndLastFrame = ctx.alloc().buffer(FRAME_HEADER_LENGTH + remainingData);
+                    flags.endOfStream(endStream);
+                    writeFrameHeaderInternal(frameHeader2AndLastFrame, remainingData, DATA, flags, streamId);
+                    // copy the last frame content
+                    frameHeader2AndLastFrame.writeBytes(data, remainingData);
+                    data = null;
```
Comment (franz1981) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:200]: Here I should release the data

**Comment by franz1981** [comment]
@ejona86 @carl-mastrangelo do you have any chance (or any pointer to make me try it) to verify/validate some of the changes I am sending recently, performance-wise? I'm thinking about gRpc over HTTP 2 and having something different from our Quarkus/Vertx stack would help a lot

**Comment by franz1981** [comment]
@bryce-anderson this is using a very similar optimization although I couldn't push it as far as http 1.1 and make it produce a single buffer overall (we don't even presize the header here :/)

**Comment by franz1981** [comment]
I am checking if is possible somehow to avoid at all to have separate buffers for headers and data, further improving things.

**Comment by carl-mastrangelo** [comment]
Alas, I have moved on from the gRPC team, and don't have access to sample http2 traffic.

**Comment by franz1981** [comment]
Thanks @carl-mastrangelo to have answered; if there is some Google users interested into grpc Netty, please reach me out here

**Comment by normanmaurer** [comment]
@franz1981 Please let me know once this is ready again for review

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 2]
```diff
@@ -135,8 +139,7 @@ public void close() { }
     @Override
     public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
             int padding, boolean endStream, ChannelPromise promise) {
-        final SimpleChannelPromiseAggregator promiseAggregator =
-                new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        SimpleChannelPromiseAggregator promiseAggregator = null;
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:142]: yeah maybe we should do this just for the sake of making it easier to read

**Comment by franz1981** [comment]
Yep, @normanmaurer this should be ready to go as well

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 3]
```diff
@@ -168,16 +174,51 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
...
-                writeFrameHeaderInternal(frameHeader2, remainingData, DATA, flags, streamId);
-                ctx.write(frameHeader2, promiseAggregator.newPromise());
-
-                // Write the payload.
-                ByteBuf lastFrame = data.readSlice(remainingData);
-                data = null;
-                ctx.write(lastFrame, promiseAggregator.newPromise());
+                // if data is an heap buffer it should still requires Netty to copy it into a direct buffer
+                // before sending it to the socket, so let's perform this copy earlier
+                if (!data.hasArray() && remainingData > COPY_SLICE_CUTOFF_BYTES) {
```
Comment (franz1981) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:179]: @bryce-anderson for heap buffers, differently from http 1.1 we don't risk to wrongly estimate the buffer, so it's safe to anticipate the copy

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 4]
```diff
@@ -498,15 +543,74 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
...
+
+    private ChannelPromise writeHeadersPartsInternal(ChannelHandlerContext ctx, int streamId, int padding,
+                                                     boolean hasPriority, int streamDependency, short weight,
+                                                     boolean exclusive, ChannelPromise promise, ByteBuf headerBlock,
+                                                     int fragmentBytes, Http2Flags flags, int nonFragmentBytes) {
+        final SimpleChannelPromiseAggregator promiseAggregator =
+                new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        try {
+            // Read the first fragment (possibly everything).
+            ByteBuf fragment = headerBlock.readRetainedSlice(fragmentBytes);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:613]: We should ensure we release this one in the case of an error otherwise we might leak

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 4 above)
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:613]: I see that this was missing in the original code as well, so I could fix it separately or contextual to this change if you wish

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:613]: maybe do a PR for the this first and then let us merge it and rebase this one . WDYT ?

Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:613]: Yep I think is a good idea, mixing up them will loose the fix, which is not nice :+1:

Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:613]: And I was wrong, it is handled, but just once :( 
I see how to fix this in this PR at this point, given that I already have to better reorganize the code...

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 5]
```diff
@@ -456,24 +520,31 @@ public ChannelFuture writeWindowUpdate(ChannelHandlerContext ctx, int streamId,
     @Override
     public ChannelFuture writeFrame(ChannelHandlerContext ctx, byte frameType, int streamId,
```
Comment (franz1981) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:521]: @normanmaurer TBH I found a LOT of place where this optmization could be applied, but the ones where I place it are the ones which popup in my profiling data. 

Maybe  @bryce-anderson or @idelpivnitskiy could spot other important points which could be worthy to fix, because still hot for other common cases

eg everytime we allocate just a  bunch of bytes and we expect to write small payloads, actually

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 6]
```diff
@@ -482,12 +553,36 @@ public ChannelFuture writeFrame(ChannelHandlerContext ctx, byte frameType, int s
         return promiseAggregator.doneAllocatingPromises();
     }
 
+    private static ChannelPromise handleFailure(ByteBuf first, ByteBuf second,
+                                                ChannelPromise promise, Throwable t,
+                                                SimpleChannelPromiseAggregator promiseAggregator) {
+        try {
+            if (first != null) {
+                first.release();
```
Comment (franz1981) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:561]: @normanmaurer  I miss Java 7 `Throwable::addSuppressed`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 7]
```diff
@@ -498,15 +593,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
...
+            }
 
             Http2Flags flags =
                     new Http2Flags().endOfStream(endStream).priorityPresent(hasPriority).paddingPresent(padding > 0);
-
-            // Read the first fragment (possibly everything).
             int nonFragmentBytes = padding + flags.getNumPriorityBytes();
             int maxFragmentLength = maxFrameSize - nonFragmentBytes;
-            ByteBuf fragment = headerBlock.readRetainedSlice(min(headerBlock.readableBytes(), maxFragmentLength));
+            int fragmentBytes = min(headerBlock.readableBytes(), maxFragmentLength);
```
Comment (franz1981) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:606]: Here, instead, I've chosen to inline the fast-path and delegate to a separate method the slowest one, mostly because It seems way more common to hit it (binary headers size below <= 128 bytes seems pretty common to me).

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 8]
```diff
@@ -135,123 +139,183 @@ public void close() { }
...
             verifyStreamId(streamId, STREAM_ID);
             verifyPadding(padding);
+        } catch (Throwable t) {
+            return handleFailure(data, null, promise, t, null);
+        }
+        if (padding == 0) {
+            return writeDataNoPadding(ctx, streamId, data, endStream, promise);
+        } else {
+            return writeDataWithPadding(ctx, streamId, data, padding, endStream, promise);
+        }
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:152]:
```suggestion
        }
        return writeDataWithPadding(ctx, streamId, data, padding, endStream, promise);
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 9]
```diff
@@ -135,123 +139,183 @@ public void close() { }
...
+                ByteBuf frameHeader2 = ctx.alloc().buffer(DATA_FRAME_HEADER_LENGTH);
+                flags.endOfStream(endStream && remainingData == 0 && padding == 0);
+                flags.paddingPresent(framePaddingBytes > 0);
+                writeFrameHeaderInternal(frameHeader2, framePaddingBytes + frameDataBytes, DATA, flags,
+                                         streamId);
+                writePaddingLength(frameHeader2, framePaddingBytes);
+                ctx.write(frameHeader2, promiseAggregator.newPromise());
+
+                // Write the payload.
+                if (data != null) { // Make sure Data is not null
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:212]:
```suggestion
                if (data != null) { // Make sure data is not null
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 10]
```diff
@@ -135,123 +139,183 @@ public void close() { }
...
-            // Use a try/finally here in case the data has been released before calling this method. This is not
-            // necessary above because we internally allocate frameHeader.
+            // Write the header.
+            final ByteBuf frameHeader2 = ctx.alloc().buffer(FRAME_HEADER_LENGTH);
+            flags.endOfStream(endStream);
+            writeFrameHeaderInternal(frameHeader2, remainingData, DATA, flags, streamId);
+            ctx.write(frameHeader2, promiseAggregator.newPromise());
+
+            // Write the payload.
+            ByteBuf lastFrame = data.readSlice(remainingData);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:267]: We could also just adjust indexes and so reduce object creation. That said maybe we should just keep it as it is as it makes things a bit easier.
```suggestion
            ByteBuf lastFrame = data.writerIndex(data.readerIndex + remainingData);
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 6 above)
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:561]: Just use `ThrowableUtils.addSuppressed(...)` ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 11]
```diff
@@ -498,15 +593,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
...
+                    buf.writeBytes(headerBlock, fragmentBytes);
+                    if (paddingBytes > 0) {
+                        // this is quite fast assuming small padding
+                        // see https://github.com/netty/netty/pull/13693
+                        buf.writeZero(paddingBytes);
+                    }
+                    ctx.write(buf, promise);
+                } catch (Throwable t) {
+                    promise.setFailure(t);
+                    PlatformDependent.throwException(t);
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:636]: why re throw ? feels wrong

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 11 above)
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:636]: Goog point; I've blindly copied from the original code this one iirc, but as you said, doesn't look right 👍

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 10 above)
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:267]: @normanmaurer I've accepted the changes!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 11 above)
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:636]: I don't feel yet comfortable to change it yet, i see that throwing the exception allow to correctly release in the caller the expected "leaky" data. I would leave it as it is, unless you want me to address the original code behaviour now (which is fine, just need to know you opinion here)

**Review by bryce-anderson — COMMENTED** [review]
Just a few comments that crossed my mind, I haven't fully looked through this yet.
The `PromiseAggregator` instances got me wondering: could this same goal be achieved using a `CompositeByteBuf`? Perhaps that is just passing the buck so to speak, but if the transport implementations are intelligently handling those types of buffers we can avoid the pipeline traversals easily.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 12]
```diff
@@ -498,15 +595,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
...
+                    if (hasPriority) {
+                        buf.writeInt(exclusive? (int) (0x80000000L | streamDependency) : streamDependency);
+                        // Adjust the weight so that it fits into a single byte on the wire.
+                        buf.writeByte(weight - 1);
+                    }
+                    buf.writeBytes(headerBlock, fragmentBytes);
+                    if (paddingBytes > 0) {
+                        // this is quite fast assuming small padding
+                        // see https://github.com/netty/netty/pull/13693
+                        buf.writeZero(paddingBytes);
```
[the remark this answers is not in the document]
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:633]: Yep, given that is very tiny implementation detail, I assumed to leverage on it just on the hottest path ie single buffer, no promise aggregation

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 13]
```diff
@@ -498,15 +595,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
 
             // Encode the entire header block.
             headerBlock = ctx.alloc().buffer();
-            headersEncoder.encodeHeaders(streamId, headers, headerBlock);
+            try {
+                headersEncoder.encodeHeaders(streamId, headers, headerBlock);
```
[the remark this answers is not in the document]
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:599]: > sincewe can determine the length of the header up front what if we reserved enough room up front for the frame header

I am not sure about it, or better, I thought we could estimate it as well (without using previous accumulated statistics as we do in http 1.1) but I couldn't find an efficient and straightforward method to do it

**Comment by franz1981** [comment]
@bryce-anderson 

> could this same goal be achieved using a CompositeByteBuf

That will both help with traversal, saving aggregate promises and saving entries on the outbound buffer, which is something I would explore regardless but, I still believe it tries to leverage writev on the transport, which is a shame to me (@normanmaurer am I right? ) when components are such small. 
To solve/improve this we could think (unless already exists, but I don't think so) about an opt-in mechanism on the transport to avoid using writev with many small sized buffers (could be either from composite or others buffer types) cause is not as efficient as it could.

While what this pr deliver is another additional improvement (without introducing new buffer types); it immediately release the buffer which contains the data, embedding its content, making it available again to be re-used and saving the inherent additional heap cost of allocating the slice, amortizing it in the buffer which embedd it.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (bryce-anderson) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:599]: Sorry, I think I shouldn't have been more clear because the term `header` is so overloaded in this context. What I meant was maybe we can skip ahead 9 bytes in `headerBlock` to reserve room for the _frame_ header and then encode the http headers to the same buffer and then go back and write the frame header. That will put the entire first HEADERS frame in the same buffer without having to do another copy regardless of the length of the encoded headers.

[the remark this answers is not in the document]
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:599]: Let me give it a shot! And it seems a good idea to me!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java [hunk 14]
```diff
@@ -482,12 +553,38 @@ public ChannelFuture writeFrame(ChannelHandlerContext ctx, byte frameType, int s
         return promiseAggregator.doneAllocatingPromises();
     }
 
+    private static ChannelPromise handleFailure(ByteBuf first, ByteBuf second,
+                                                ChannelPromise promise, Throwable t,
+                                                SimpleChannelPromiseAggregator promiseAggregator) {
```
[the remark this answers is not in the document]
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:561]: Actually if one is specified it takes precedence: I'm adding a comment in the code for it (or I should enforce that `promise` is `null` if the aggregator is not)

At aab16822d4f92b52312c2e0c93baf61ff65c37b4

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (franz1981) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java:619]: Addressed at aab16822d4f92b52312c2e0c93baf61ff65c37b4

The heuristic here works like that:
- if the padding bytes are less then 64, we can zero efficiently using `writeZero` 
- if we can zero efficiently, let's try skipping the right amount of frame header bytes directly in the header block
- if we have been lucky, we could reuse directly the available space within header block OR (ie no luck) if the bytes to copy from it are below 128, we could allocate a right sized one

I can decide to be more strict and encode the "shifted" headers only if padding == 0, but it seems too strict to me; we can be really be lucky with `256 bytes - HEADERS_FRAME_HEADER_LENGTH = 241` bytes available for headers.

**Comment by franz1981** [comment]
PTAL @bryce-anderson @normanmaurer

**Comment by franz1981** [comment]
@normanmaurer I see that the original code was throwing exceptions, while not expected AND not correctly releasing multiple retained header buffers; I could try fix the former but the latter would requires a separate pr really, given that seems a much more annoying and complex technical debt, which didn't yet bitten us.

**Comment by franz1981** [comment]
mmm @normanmaurer what happened for 33f5c561ff17a4e67e3d94b27292a5f0bc304a49? :P

**Comment by franz1981** [comment]
@normanmaurer I've received notification of `Merge branch '4.1' into 4.1_no_slice` changes, but IDK what it is :P

**Comment by normanmaurer** [comment]
@franz1981 ah this was just to bring it up to date with current 4.1

**Comment by franz1981** [comment]
@normanmaurer there is something more you want me to take a look for this?

As said, I am not very confident the original code was correctly handling leaks here, but is really tons of code and we should address it separately (eg all the header retained slices around, the thrown you noticed in some earlier comment) - ideally first, if we think them to be critic.

What I could do is to try address them already for the fast-paths I've introduced, but will create the weird situation, where the optimized (although common enough, I suppose) new stuff is more correct than what exists from some time already...

**Comment by normanmaurer** [comment]
@franz1981 just hold of a bit... I will try to find time to review this one first

**Comment by franz1981** [comment]
@bryce-anderson if you want to give it another shot, I have tried implementing your suggestion to further reduce the number of allocations

**Comment by franz1981** [comment]
Any other concerns @bryce-anderson or @normanmaurer ? 
I can try reducing the amount of changes if it helps

**Comment by franz1981** [comment]
@normanmaurer @chrisvest any news for this folks?
Not urgent but the performance effects was rather positives....

**Comment by normanmaurer** [comment]
@ejona86 @idelpivnitskiy @bryce-anderson can you check again ?

**Comment by franz1981** [comment]
As soon as I got from PTO i would like to proceed on this... @bryce-anderson @ejona86 @idelpivnitskiy wdyt about the approach? Any other concern?

**Comment by normanmaurer** [comment]
sure lets do it

**Comment by normanmaurer** [comment]
@franz1981 do we want to rebase and get this in ?

**Comment by franz1981** [comment]
Try it if you can @normanmaurer it was a good improvement at the time I have done it

**Comment by chrisvest** [comment]
I tried cherry-picking this to 4.2, but it doesn't look beneficial on the face of it:

Current 4.2:

```
Benchmark                                (padding)  (payloadSize)  (pooled)  Mode  Cnt    Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter          0             64      true  avgt    5   64.933 ± 0.610  ns/op
Http2FrameWriterDataBenchmark.newWriter          0             64     false  avgt    5  318.025 ± 3.257  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024      true  avgt    5   60.911 ± 0.777  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024     false  avgt    5  309.507 ± 3.505  ns/op
Benchmark                                (padding)  (payloadSize)  (pooled)  Mode  Cnt    Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter          0             64      true  avgt    5   67.270 ± 6.872  ns/op
Http2FrameWriterDataBenchmark.newWriter          0             64     false  avgt    5  310.112 ± 4.378  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024      true  avgt    5   60.223 ± 1.268  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024     false  avgt    5  306.737 ± 4.231  ns/op
```


This PR cherry-picked:

```
Benchmark                                (padding)  (payloadSize)  (pooled)  Mode  Cnt    Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter          0             64      true  avgt    5   71.655 ± 0.913  ns/op
Http2FrameWriterDataBenchmark.newWriter          0             64     false  avgt    5  303.061 ± 5.332  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024      true  avgt    5   60.831 ± 1.914  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024     false  avgt    5  304.938 ± 7.290  ns/op
Benchmark                                (padding)  (payloadSize)  (pooled)  Mode  Cnt    Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter          0             64      true  avgt    5   69.571 ± 0.767  ns/op
Http2FrameWriterDataBenchmark.newWriter          0             64     false  avgt    5  300.887 ± 5.980  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024      true  avgt    5   64.977 ± 2.050  ns/op
Http2FrameWriterDataBenchmark.newWriter          0           1024     false  avgt    5  303.579 ± 3.097  ns/op
```

**Comment by franz1981** [comment]
This was beneficial with POSIX writev last time I checked it

**Comment by franz1981** [comment]
@chrisvest Running the JMH benchmark with the existing implementation  (not this PR) I see few interesting things

image

This is where I have highlighted the allocation of the promise aggregator; it's very tiny in the benchmark, so i can see why it is not that important.
More: `io/netty/channel/embedded/EmbeddedChannel.maybeRunPendingTasks` is the most dominant cost (on the right) and there's no actual syscall `writev` or whatever.
In short, the main reasons why this PR was beneficials are gone

**Comment by franz1981** [comment]
I've just tried @chrisvest 

image


`writeBytes` strikes again: the performance IS improved but this `writeBytes` make it on par, again.
I think it's something we have to think about...

**Comment by franz1981** [comment]
FYI @chrisvest I've applied a small patch to the adaptive byte buf (explained at https://github.com/netty/netty/issues/15723#issuecomment-3364355756) which shows the improvement I've mentioned (and it doesn't still show the benefits re syscalls `writev` vs `send`!):
before:
```
Benchmark                                       (padding)  (payloadSize)  (pooled)  Mode  Cnt   Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter                 0            128      true  avgt    5  63.886 ± 0.933  ns/op
```
after:
```
Benchmark                                       (padding)  (payloadSize)  (pooled)  Mode  Cnt   Score   Error  Units
Http2FrameWriterDataBenchmark.newWriter                 0            128      true  avgt    5  50.693 ± 0.823  ns/op
```
which is similar to what I was claiming at the beginning for this PR.

**Comment by franz1981** [comment]
@chrisvest I could revamp this WDYT?
I was waiting some perf changes from you which has been merged ❤️
